File: SwiGi.Win/SwiGi.Win/Hid/DeviceDiscovery.cs
using System.Runtime.InteropServices;

namespace SwiGi.Win.Hid;

internal static class DeviceDiscovery
{
    public static DeviceInfo? FindDevice(byte wantedType)
    {
        var head = HidApiNative.hid_enumerate(HidPPConstants.LogitechVid, 0);
        if (head == IntPtr.Zero) return null;

        var candidates = new List<(int Score, string Path, ushort Pid, ushort UsagePage, ushort Usage)>();

        try
        {
            var node = head;
            while (node != IntPtr.Zero)
            {
                var info = Marshal.PtrToStructure<HidApiNative.HidDeviceInfo>(node);
                node = info.Next;

                var pid = info.ProductId;
                var usagePage = info.UsagePage;
                var usage = info.Usage;

                if (HidPPConstants.AllReceiverPids.Contains(pid)) continue;
                if (!HidPPConstants.DirectUsagePairs.Contains((usagePage, usage))) continue;

                var score = usagePage is 0xFF00 or 0xFF43 or 0xFF0C ? 100 : 0;
                var path = HidApiNative.PathToString(info.Path);
                if (string.IsNullOrEmpty(path)) continue;

                candidates.Add((score, path, pid, usagePage, usage));
            }
        }
        finally
        {
            HidApiNative.hid_free_enumeration(head);
        }

        candidates.Sort((a, b) => b.Score.CompareTo(a.Score));

        var foundPids = new HashSet<ushort>();
        foreach (var (_, path, pid, _, _) in candidates)
        {
            if (foundPids.Contains(pid)) continue;

            HidTransport? transport = null;
            try
            {
                transport = new HidTransport(path, pid);

                var featureIndex = HidPPProtocol.ResolveFeature(transport, HidPPConstants.DevnumberDirect, HidPPConstants.FeatureDeviceTypeAndName);
                if (featureIndex is null)
                {
                    transport.Dispose();
                    continue;
                }

                var deviceType = HidPPProtocol.GetDeviceType(transport, HidPPConstants.DevnumberDirect, featureIndex.Value);
                if (deviceType is null)
                {
                    transport.Dispose();
                    continue;
                }

                var name = HidPPProtocol.GetDeviceName(transport, HidPPConstants.DevnumberDirect, featureIndex.Value)
                    ?? $"Logitech-0x{pid:X4}";

                var isMouse = deviceType is HidPPConstants.DeviceTypeMouse
                    or HidPPConstants.DeviceTypeTrackpad
                    or HidPPConstants.DeviceTypeTrackball;

                if (wantedType == HidPPConstants.DeviceTypeKeyboard && deviceType != HidPPConstants.DeviceTypeKeyboard)
                {
                    transport.Dispose();
                    continue;
                }

                if (wantedType == HidPPConstants.DeviceTypeMouse && !isMouse)
                {
                    transport.Dispose();
                    continue;
                }

                var changeHostIndex = HidPPProtocol.ResolveFeature(transport, HidPPConstants.DevnumberDirect, HidPPConstants.FeatureChangeHost);
                if (changeHostIndex is null)
                {
                    transport.Dispose();
                    continue;
                }

                foundPids.Add(pid);
                return new DeviceInfo(transport, name, pid, changeHostIndex.Value);
            }
            catch
            {
                transport?.Dispose();
            }
        }

        return null;
    }
}

File: SwiGi/SwiGi/DeviceDiscovery.swift
import Foundation
import CHIDAPI

struct DeviceInfo {
    let transport: HIDTransport
    let name: String
    let pid: UInt16
    let changeHostIndex: UInt8

    mutating func close() {
        transport.close()
    }
}

enum DeviceDiscovery {
    static func findDevice(wantedType: UInt8) -> DeviceInfo? {
        guard let head = hid_enumerate(HIDPPConstants.logitechVID, 0) else { return nil }
        defer { hid_free_enumeration(head) }

        var candidates: [(score: Int, path: String, pid: UInt16, usagePage: UInt16, usage: UInt16)] = []
        var node: UnsafeMutablePointer<hid_device_info>? = head

        while let current = node {
            let info = current.pointee
            node = info.next

            let pid = info.product_id
            let usagePage = info.usage_page
            let usage = info.usage

            if HIDPPConstants.allReceiverPIDs.contains(pid) { continue }
            let pairKey = HIDPPConstants.usagePairKey(page: usagePage, usage: usage)
            guard HIDPPConstants.directUsagePairs.contains(pairKey) else { continue }

            let score = [0xFF00, 0xFF43, 0xFF0C].contains(usagePage) ? 100 : 0
            let path = String(cString: info.path)
            candidates.append((score, path, pid, usagePage, usage))
        }

        candidates.sort { $0.score > $1.score }

        var foundPIDs = Set<UInt16>()
        for candidate in candidates {
            if foundPIDs.contains(candidate.pid) { continue }

            let transport: HIDTransport
            do {
                transport = try HIDTransport(path: candidate.path, pid: candidate.pid)
            } catch {
                continue
            }

            do {
                guard let featureIndex = HIDPPProtocol.resolveFeature(
                    transport: transport,
                    devnumber: HIDPPConstants.devnumberDirect,
                    featureCode: HIDPPConstants.featureDeviceTypeAndName
                ) else {
                    transport.close()
                    continue
                }

                guard let deviceType = HIDPPProtocol.getDeviceType(
                    transport: transport,
                    devnumber: HIDPPConstants.devnumberDirect,
                    featureIndex: featureIndex
                ) else {
                    transport.close()
                    continue
                }

                let name = HIDPPProtocol.getDeviceName(
                    transport: transport,
                    devnumber: HIDPPConstants.devnumberDirect,
                    featureIndex: featureIndex
                ) ?? "Logitech-0x\(String(format: "%04X", candidate.pid))"

                let isMouse = [HIDPPConstants.deviceTypeMouse, HIDPPConstants.deviceTypeTrackpad, HIDPPConstants.deviceTypeTrackball]
                    .contains(deviceType)

                if wantedType == HIDPPConstants.deviceTypeKeyboard, deviceType != HIDPPConstants.deviceTypeKeyboard {
                    transport.close()
                    continue
                }
                if wantedType == HIDPPConstants.deviceTypeMouse, !isMouse {
                    transport.close()
                    continue
                }

                guard let changeHostIndex = HIDPPProtocol.resolveFeature(
                    transport: transport,
                    devnumber: HIDPPConstants.devnumberDirect,
                    featureCode: HIDPPConstants.featureChangeHost
                ) else {
                    transport.close()
                    continue
                }

                foundPIDs.insert(candidate.pid)
                return DeviceInfo(
                    transport: transport,
                    name: name,
                    pid: candidate.pid,
                    changeHostIndex: changeHostIndex
                )
            } catch {
                transport.close()
                continue
            }
        }

        return nil
    }

    /// Lists Logitech HID interfaces visible to hidapi (for troubleshooting logs).
    static func scanSummary() -> String {
        guard let head = hid_enumerate(HIDPPConstants.logitechVID, 0) else {
            return "No Logitech HID devices visible to hidapi."
        }
        defer { hid_free_enumeration(head) }

        var lines: [String] = []
        var node: UnsafeMutablePointer<hid_device_info>? = head
        while let current = node {
            let info = current.pointee
            node = info.next
            let pid = info.product_id
            if HIDPPConstants.allReceiverPIDs.contains(pid) { continue }
            let pairKey = HIDPPConstants.usagePairKey(page: info.usage_page, usage: info.usage)
            let path = String(cString: info.path)
            let supported = HIDPPConstants.directUsagePairs.contains(pairKey) ? "ok" : "skip"
            lines.append(
                "  pid=0x\(String(format: "%04X", pid)) usage=0x\(String(format: "%04X", info.usage_page)):\(String(format: "%04X", info.usage)) [\(supported)] \(path)"
            )
        }
        if lines.isEmpty {
            return "No Logitech HID devices visible to hidapi (check Bluetooth and Input Monitoring)."
        }
        return "Logitech HID scan:\n" + lines.joined(separator: "\n")
    }
}
